File: archive/pulumi-py/Pr4300/lib/compute_stack.py
"""
compute_stack.py

Multi-region compute infrastructure with ALB and Lambda functions.
"""

import pulumi
import pulumi_aws as aws
from pulumi import Output, ResourceOptions
import json


class ComputeStack(pulumi.ComponentResource):
    """
    Creates multi-region compute infrastructure for application hosting.
    """

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        primary_region: str,
        secondary_region: str,
        primary_vpc_id: Output,
        secondary_vpc_id: Output,
        primary_subnet_ids: list,
        secondary_subnet_ids: list,
        database_endpoint: Output,
        storage_bucket: Output,
        tags: dict,
        opts: ResourceOptions = None
    ):
        super().__init__('tap:compute:ComputeStack', name, None, opts)

        secondary_provider = aws.Provider(
            f"compute-secondary-provider-{environment_suffix}",
            region=secondary_region,
            opts=ResourceOptions(parent=self)
        )

        # Security group for ALB in primary region
        self.primary_alb_sg = aws.ec2.SecurityGroup(
            f"primary-alb-sg-{environment_suffix}",
            vpc_id=primary_vpc_id,
            description="Security group for primary ALB",
            ingress=[
                aws.ec2.SecurityGroupIngressArgs(
                    protocol="tcp",
                    from_port=80,
                    to_port=80,
                    cidr_blocks=["0.0.0.0/0"],
                ),
                aws.ec2.SecurityGroupIngressArgs(
                    protocol="tcp",
                    from_port=443,
                    to_port=443,
                    cidr_blocks=["0.0.0.0/0"],
                )
            ],
            egress=[
                aws.ec2.SecurityGroupEgressArgs(
                    protocol="-1",
                    from_port=0,
                    to_port=0,
                    cidr_blocks=["0.0.0.0/0"],
                )
            ],
            tags={**tags, 'Name': f'primary-alb-sg-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        # Security group for ALB in secondary region
        self.secondary_alb_sg = aws.ec2.SecurityGroup(
            f"secondary-alb-sg-{environment_suffix}",
            vpc_id=secondary_vpc_id,
            description="Security group for secondary ALB",
            ingress=[
                aws.ec2.SecurityGroupIngressArgs(
                    protocol="tcp",
                    from_port=80,
                    to_port=80,
                    cidr_blocks=["0.0.0.0/0"],
                ),
                aws.ec2.SecurityGroupIngressArgs(
                    protocol="tcp",
                    from_port=443,
                    to_port=443,
                    cidr_blocks=["0.0.0.0/0"],
                )
            ],
            egress=[
                aws.ec2.SecurityGroupEgressArgs(
                    protocol="-1",
                    from_port=0,
                    to_port=0,
                    cidr_blocks=["0.0.0.0/0"],
                )
            ],
            tags={**tags, 'Name': f'secondary-alb-sg-{environment_suffix}'},
            opts=ResourceOptions(parent=self, provider=secondary_provider)
        )

        # Primary ALB
        self.primary_alb = aws.lb.LoadBalancer(
            f"primary-alb-{environment_suffix}",
            name=f"primary-alb-{environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[self.primary_alb_sg.id],
            subnets=primary_subnet_ids,
            enable_deletion_protection=False,
            tags={**tags, 'Name': f'primary-alb-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        # Secondary ALB
        self.secondary_alb = aws.lb.LoadBalancer(
            f"secondary-alb-{environment_suffix}",
            name=f"secondary-alb-{environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[self.secondary_alb_sg.id],
            subnets=secondary_subnet_ids,
            enable_deletion_protection=False,
            tags={**tags, 'Name': f'secondary-alb-{environment_suffix}'},
            opts=ResourceOptions(parent=self, provider=secondary_provider)
        )

        # Target groups
        self.primary_target_group = aws.lb.TargetGroup(
            f"primary-tg-{environment_suffix}",
            name=f"primary-tg-{environment_suffix}",
            port=80,
            protocol="HTTP",
            vpc_id=primary_vpc_id,
            target_type="lambda",
            health_check=aws.lb.TargetGroupHealthCheckArgs(
                enabled=True,
                path="/health",
                interval=30,
                timeout=5,
                healthy_threshold=2,
                unhealthy_threshold=2,
            ),
            tags={**tags, 'Name': f'primary-tg-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        self.secondary_target_group = aws.lb.TargetGroup(
            f"secondary-tg-{environment_suffix}",
            name=f"secondary-tg-{environment_suffix}",
            port=80,
            protocol="HTTP",
            vpc_id=secondary_vpc_id,
            target_type="lambda",
            health_check=aws.lb.TargetGroupHealthCheckArgs(
                enabled=True,
                path="/health",
                interval=30,
                timeout=5,
                healthy_threshold=2,
                unhealthy_threshold=2,
            ),
            tags={**tags, 'Name': f'secondary-tg-{environment_suffix}'},
            opts=ResourceOptions(parent=self, provider=secondary_provider)
        )

        # Listeners
        self.primary_listener = aws.lb.Listener(
            f"primary-listener-{environment_suffix}",
            load_balancer_arn=self.primary_alb.arn,
            port=80,
            protocol="HTTP",
            default_actions=[aws.lb.ListenerDefaultActionArgs(
                type="forward",
                target_group_arn=self.primary_target_group.arn,
            )],
            tags={**tags, 'Name': f'primary-listener-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        self.secondary_listener = aws.lb.Listener(
            f"secondary-listener-{environment_suffix}",
            load_balancer_arn=self.secondary_alb.arn,
            port=80,
            protocol="HTTP",
            default_actions=[aws.lb.ListenerDefaultActionArgs(
                type="forward",
                target_group_arn=self.secondary_target_group.arn,
            )],
            tags={**tags, 'Name': f'secondary-listener-{environment_suffix}'},
            opts=ResourceOptions(parent=self, provider=secondary_provider)
        )

        # IAM role for Lambda
        lambda_role = aws.iam.Role(
            f"lambda-exec-role-{environment_suffix}",
            assume_role_policy=json.dumps({
                "Version": "2012-10-17",
                "Statement": [{
                    "Effect": "Allow",
                    "Principal": {"Service": "lambda.amazonaws.com"},
                    "Action": "sts:AssumeRole"
                }]
            }),
            managed_policy_arns=[
                "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
                "arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
            ],
            tags={**tags, 'Name': f'lambda-exec-role-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )
        # Lambda function for primary region
        self.primary_lambda = aws.lambda_.Function(
            f"primary-app-lambda-{environment_suffix}",
            name=f"primary-app-lambda-{environment_suffix}",
            runtime="python3.11",
            role=lambda_role.arn,
            handler="index.handler",
            code=pulumi.AssetArchive({
                "index.py": pulumi.StringAsset("""
import json

def handler(event, context):
    return {
        'statusCode': 200,
        'headers': {'Content-Type': 'application/json'},
        'body': json.dumps({
            'message': 'E-commerce API - Primary Region',
            'region': 'us-east-1',
            'status': 'active'
        })
    }
""")
            }),
            timeout=30,
            memory_size=256,
            environment=aws.lambda_.FunctionEnvironmentArgs(
                variables={
                    "REGION": primary_region,
                    "DB_ENDPOINT": database_endpoint,
                    "STORAGE_BUCKET": storage_bucket,
                }
            ),
            tags={**tags, 'Name': f'primary-app-lambda-{environment_suffix}'},
            opts=ResourceOptions(parent=self, depends_on=[lambda_role])
        )

        # Lambda function for secondary region
        self.secondary_lambda = aws.lambda_.Function(
            f"secondary-app-lambda-{environment_suffix}",
            name=f"secondary-app-lambda-{environment_suffix}",
            runtime="python3.11",
            role=lambda_role.arn,
            handler="index.handler",
            code=pulumi.AssetArchive({
                "index.py": pulumi.StringAsset("""
import json

def handler(event, context):
    return {
        'statusCode': 200,
        'headers': {'Content-Type': 'application/json'},
        'body': json.dumps({
            'message': 'E-commerce API - Secondary Region',
            'region': 'us-west-2',
            'status': 'standby'
        })
    }
""")
            }),
            timeout=30,
            memory_size=256,
            environment=aws.lambda_.FunctionEnvironmentArgs(
                variables={
                    "REGION": secondary_region,
                }
            ),
            tags={**tags, 'Name': f'secondary-app-lambda-{environment_suffix}'},
            opts=ResourceOptions(parent=self, provider=secondary_provider, depends_on=[lambda_role])
        )

        # Lambda permissions for ALB
        primary_lambda_permission = aws.lambda_.Permission(
            f"primary-lambda-alb-permission-{environment_suffix}",
            action="lambda:InvokeFunction",
            function=self.primary_lambda.name,
            principal="elasticloadbalancing.amazonaws.com",
            source_arn=self.primary_target_group.arn,
            opts=ResourceOptions(parent=self)
        )

        secondary_lambda_permission = aws.lambda_.Permission(
            f"secondary-lambda-alb-permission-{environment_suffix}",
            action="lambda:InvokeFunction",
            function=self.secondary_lambda.name,
            principal="elasticloadbalancing.amazonaws.com",
            source_arn=self.secondary_target_group.arn,
            opts=ResourceOptions(parent=self, provider=secondary_provider)
        )

        # Attach Lambda to target groups
        # Must depend on permission to ensure ELB can invoke Lambda
        aws.lb.TargetGroupAttachment(
            f"primary-lambda-tg-attachment-{environment_suffix}",
            target_group_arn=self.primary_target_group.arn,
            target_id=self.primary_lambda.arn,
            opts=ResourceOptions(parent=self, depends_on=[self.primary_lambda, primary_lambda_permission])
        )

        aws.lb.TargetGroupAttachment(
            f"secondary-lambda-tg-attachment-{environment_suffix}",
            target_group_arn=self.secondary_target_group.arn,
            target_id=self.secondary_lambda.arn,
            opts=ResourceOptions(
                parent=self,
                provider=secondary_provider,
                depends_on=[self.secondary_lambda, secondary_lambda_permission]
            )
        )

        # Expose outputs
        self.primary_alb_dns = self.primary_alb.dns_name
        self.secondary_alb_dns = self.secondary_alb.dns_name
        self.primary_target_group_arn = self.primary_target_group.arn
        self.primary_lambda_arn = self.primary_lambda.arn
        self.secondary_lambda_arn = self.secondary_lambda.arn

        self.register_outputs({})

File: archive/pulumi-ts/Pr6318/lib/compute-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';
import * as random from '@pulumi/random';

export interface ComputeStackArgs {
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
  primaryVpcId: pulumi.Output<string>;
  drVpcId: pulumi.Output<string>;
  primaryPublicSubnetIds: pulumi.Output<string[]>;
  drPublicSubnetIds: pulumi.Output<string[]>;
  primaryPrivateSubnetIds: pulumi.Output<string[]>;
  drPrivateSubnetIds: pulumi.Output<string[]>;
  primaryProvider: aws.Provider;
  drProvider: aws.Provider;
}

export class ComputeStack extends pulumi.ComponentResource {
  public readonly primaryAlbDnsName: pulumi.Output<string>;
  public readonly drAlbDnsName: pulumi.Output<string>;
  public readonly primaryAlbZoneId: pulumi.Output<string>;
  public readonly drAlbZoneId: pulumi.Output<string>;
  public readonly primaryAlbArn: pulumi.Output<string>;
  public readonly drAlbArn: pulumi.Output<string>;
  public readonly primaryInstanceSecurityGroupId: pulumi.Output<string>;
  public readonly drInstanceSecurityGroupId: pulumi.Output<string>;

  constructor(
    name: string,
    args: ComputeStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:compute:ComputeStack', name, args, opts);

    const {
      environmentSuffix,
      tags,
      primaryVpcId,
      drVpcId,
      primaryPublicSubnetIds,
      drPublicSubnetIds,
      primaryPrivateSubnetIds,
      drPrivateSubnetIds,
      primaryProvider,
      drProvider,
    } = args;

    // Skip HTTPS for test/PR environments to avoid ACM certificate validation timeouts
    const enableHttps = !environmentSuffix.toLowerCase().startsWith('pr');

    // Generate random suffix to avoid resource name conflicts
    const randomSuffix = new random.RandomString(
      `compute-random-suffix-${environmentSuffix}`,
      {
        length: 8,
        special: false,
        upper: false,
        lower: true,
        numeric: true,
      },
      { parent: this }
    );

    // Get latest Amazon Linux 2 AMI for primary region
    const primaryAmi = aws.ec2.getAmi(
      {
        mostRecent: true,
        owners: ['amazon'],
        filters: [
          {
            name: 'name',
            values: ['amzn2-ami-hvm-*-x86_64-gp2'],
          },
        ],
      },
      { provider: primaryProvider }
    );

    // Get latest Amazon Linux 2 AMI for DR region
    const drAmi = aws.ec2.getAmi(
      {
        mostRecent: true,
        owners: ['amazon'],
        filters: [
          {
            name: 'name',
            values: ['amzn2-ami-hvm-*-x86_64-gp2'],
          },
        ],
      },
      { provider: drProvider }
    );

    // Security group for ALB in primary region
    const primaryAlbIngressRules = [
      {
        protocol: 'tcp',
        fromPort: 80,
        toPort: 80,
        cidrBlocks: ['0.0.0.0/0'],
        description: 'Allow HTTP traffic from internet',
      },
    ];

    if (enableHttps) {
      primaryAlbIngressRules.push({
        protocol: 'tcp',
        fromPort: 443,
        toPort: 443,
        cidrBlocks: ['0.0.0.0/0'],
        description: 'Allow HTTPS traffic from internet',
      });
    }

    const primaryAlbSg = new aws.ec2.SecurityGroup(
      `primary-alb-sg-${environmentSuffix}`,
      {
        vpcId: primaryVpcId,
        description: 'Security group for primary ALB',
        ingress: primaryAlbIngressRules,
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
          },
        ],
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `primary-alb-sg-${environmentSuffix}`,
          'DR-Role': 'primary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Security group for ALB in DR region
    const drAlbIngressRules = [
      {
        protocol: 'tcp',
        fromPort: 80,
        toPort: 80,
        cidrBlocks: ['0.0.0.0/0'],
        description: 'Allow HTTP traffic from internet',
      },
    ];

    if (enableHttps) {
      drAlbIngressRules.push({
        protocol: 'tcp',
        fromPort: 443,
        toPort: 443,
        cidrBlocks: ['0.0.0.0/0'],
        description: 'Allow HTTPS traffic from internet',
      });
    }

    const drAlbSg = new aws.ec2.SecurityGroup(
      `dr-alb-sg-${environmentSuffix}`,
      {
        vpcId: drVpcId,
        description: 'Security group for DR ALB',
        ingress: drAlbIngressRules,
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
          },
        ],
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `dr-alb-sg-${environmentSuffix}`,
          'DR-Role': 'secondary',
        })),
      },
      { provider: drProvider, parent: this }
    );

    // Security group for EC2 instances in primary region
    const primaryInstanceSg = new aws.ec2.SecurityGroup(
      `primary-instance-sg-${environmentSuffix}`,
      {
        vpcId: primaryVpcId,
        description: 'Security group for primary EC2 instances',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            securityGroups: [primaryAlbSg.id],
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
          },
        ],
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `primary-instance-sg-${environmentSuffix}`,
          'DR-Role': 'primary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Security group for EC2 instances in DR region
    const drInstanceSg = new aws.ec2.SecurityGroup(
      `dr-instance-sg-${environmentSuffix}`,
      {
        vpcId: drVpcId,
        description: 'Security group for DR EC2 instances',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            securityGroups: [drAlbSg.id],
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
          },
        ],
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `dr-instance-sg-${environmentSuffix}`,
          'DR-Role': 'secondary',
        })),
      },
      { provider: drProvider, parent: this }
    );

    // User data script with health endpoint
    const userData = `#!/bin/bash
yum update -y
yum install -y httpd
systemctl start httpd
systemctl enable httpd

# Main page
REGION=$(ec2-metadata --availability-zone | cut -d' ' -f2)
echo "<h1>Payment Processing System</h1><p>Region: $REGION</p><p>Status: Active</p>" > /var/www/html/index.html

# Health endpoint for Route53 health checks
echo "OK" > /var/www/html/health.html
`;

    // Launch template for primary region
    const primaryLaunchTemplate = new aws.ec2.LaunchTemplate(
      `primary-lt-${environmentSuffix}`,
      {
        namePrefix: pulumi.interpolate`primary-lt-${environmentSuffix}-${randomSuffix.result}-`,
        imageId: pulumi.output(primaryAmi).apply(ami => ami.id),
        instanceType: 't3.medium',
        vpcSecurityGroupIds: [primaryInstanceSg.id],
        userData: Buffer.from(userData).toString('base64'),
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `primary-lt-${environmentSuffix}`,
          'DR-Role': 'primary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Launch template for DR region
    const drLaunchTemplate = new aws.ec2.LaunchTemplate(
      `dr-lt-${environmentSuffix}`,
      {
        namePrefix: pulumi.interpolate`dr-lt-${environmentSuffix}-${randomSuffix.result}-`,
        imageId: pulumi.output(drAmi).apply(ami => ami.id),
        instanceType: 't3.medium',
        vpcSecurityGroupIds: [drInstanceSg.id],
        userData: Buffer.from(userData).toString('base64'),
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `dr-lt-${environmentSuffix}`,
          'DR-Role': 'secondary',
        })),
      },
      { provider: drProvider, parent: this }
    );

    // Target group for primary ALB
    const primaryTargetGroup = new aws.lb.TargetGroup(
      `primary-tg-${environmentSuffix}`,
      {
        name: pulumi.interpolate`primary-tg-${environmentSuffix}-${randomSuffix.result}`,
        port: 80,
        protocol: 'HTTP',
        vpcId: primaryVpcId,
        healthCheck: {
          enabled: true,
          path: '/health.html',
          protocol: 'HTTP',
          matcher: '200',
          interval: 30,
          timeout: 5,
          healthyThreshold: 2,
          unhealthyThreshold: 2,
        },
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `primary-tg-${environmentSuffix}`,
          'DR-Role': 'primary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Target group for DR ALB
    const drTargetGroup = new aws.lb.TargetGroup(
      `dr-tg-${environmentSuffix}`,
      {
        name: pulumi.interpolate`dr-tg-${environmentSuffix}-${randomSuffix.result}`,
        port: 80,
        protocol: 'HTTP',
        vpcId: drVpcId,
        healthCheck: {
          enabled: true,
          path: '/health.html',
          protocol: 'HTTP',
          matcher: '200',
          interval: 30,
          timeout: 5,
          healthyThreshold: 2,
          unhealthyThreshold: 2,
        },
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `dr-tg-${environmentSuffix}`,
          'DR-Role': 'secondary',
        })),
      },
      { provider: drProvider, parent: this }
    );

    // Auto Scaling Group for primary region
    const _primaryAsg = new aws.autoscaling.Group(
      `primary-asg-${environmentSuffix}`,
      {
        name: pulumi.interpolate`primary-asg-${environmentSuffix}-${randomSuffix.result}`,
        vpcZoneIdentifiers: primaryPrivateSubnetIds,
        targetGroupArns: [primaryTargetGroup.arn],
        minSize: 2,
        maxSize: 6,
        desiredCapacity: 2,
        healthCheckType: 'ELB',
        healthCheckGracePeriod: 300,
        launchTemplate: {
          id: primaryLaunchTemplate.id,
          version: '$Latest',
        },
        tags: [
          {
            key: 'Name',
            value: `primary-asg-instance-${environmentSuffix}`,
            propagateAtLaunch: true,
          },
          {
            key: 'DR-Role',
            value: 'primary',
            propagateAtLaunch: true,
          },
          {
            key: 'Environment',
            value: environmentSuffix,
            propagateAtLaunch: true,
          },
        ],
      },
      { provider: primaryProvider, parent: this }
    );
    void _primaryAsg;

    // Auto Scaling Group for DR region
    const _drAsg = new aws.autoscaling.Group(
      `dr-asg-${environmentSuffix}`,
      {
        name: pulumi.interpolate`dr-asg-${environmentSuffix}-${randomSuffix.result}`,
        vpcZoneIdentifiers: drPrivateSubnetIds,
        targetGroupArns: [drTargetGroup.arn],
        minSize: 2,
        maxSize: 6,
        desiredCapacity: 2,
        healthCheckType: 'ELB',
        healthCheckGracePeriod: 300,
        launchTemplate: {
          id: drLaunchTemplate.id,
          version: '$Latest',
        },
        tags: [
          {
            key: 'Name',
            value: `dr-asg-instance-${environmentSuffix}`,
            propagateAtLaunch: true,
          },
          {
            key: 'DR-Role',
            value: 'secondary',
            propagateAtLaunch: true,
          },
          {
            key: 'Environment',
            value: environmentSuffix,
            propagateAtLaunch: true,
          },
        ],
      },
      { provider: drProvider, parent: this }
    );
    void _drAsg;

    // Application Load Balancer for primary region
    const primaryAlb = new aws.lb.LoadBalancer(
      `primary-alb-${environmentSuffix}`,
      {
        name: pulumi.interpolate`primary-alb-${environmentSuffix}-${randomSuffix.result}`,
        internal: false,
        loadBalancerType: 'application',
        securityGroups: [primaryAlbSg.id],
        subnets: primaryPublicSubnetIds,
        enableHttp2: true,
        enableDeletionProtection: false,
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `primary-alb-${environmentSuffix}`,
          'DR-Role': 'primary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Application Load Balancer for DR region
    const drAlb = new aws.lb.LoadBalancer(
      `dr-alb-${environmentSuffix}`,
      {
        name: pulumi.interpolate`dr-alb-${environmentSuffix}-${randomSuffix.result}`,
        internal: false,
        loadBalancerType: 'application',
        securityGroups: [drAlbSg.id],
        subnets: drPublicSubnetIds,
        enableHttp2: true,
        enableDeletionProtection: false,
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `dr-alb-${environmentSuffix}`,
          'DR-Role': 'secondary',
        })),
      },
      { provider: drProvider, parent: this }
    );

    // HTTPS setup (only for production environments)
    if (enableHttps) {
      // ACM Certificate for primary region
      // NOTE: In production, replace with a validated certificate for your domain
      const primaryCert = new aws.acm.Certificate(
        `primary-cert-${environmentSuffix}`,
        {
          domainName: `payments-${environmentSuffix}.example.com`,
          validationMethod: 'DNS',
          subjectAlternativeNames: [
            `*.payments-${environmentSuffix}.example.com`,
          ],
          tags: pulumi.all([tags]).apply(([t]) => ({
            ...t,
            Name: `primary-cert-${environmentSuffix}`,
            'DR-Role': 'primary',
          })),
        },
        { provider: primaryProvider, parent: this }
      );

      // ACM Certificate for DR region
      // NOTE: In production, replace with a validated certificate for your domain
      const drCert = new aws.acm.Certificate(
        `dr-cert-${environmentSuffix}`,
        {
          domainName: `payments-${environmentSuffix}.example.com`,
          validationMethod: 'DNS',
          subjectAlternativeNames: [
            `*.payments-${environmentSuffix}.example.com`,
          ],
          tags: pulumi.all([tags]).apply(([t]) => ({
            ...t,
            Name: `dr-cert-${environmentSuffix}`,
            'DR-Role': 'secondary',
          })),
        },
        { provider: drProvider, parent: this }
      );

      // HTTPS Listener for primary ALB
      new aws.lb.Listener(
        `primary-alb-https-listener-${environmentSuffix}`,
        {
          loadBalancerArn: primaryAlb.arn,
          port: 443,
          protocol: 'HTTPS',
          sslPolicy: 'ELBSecurityPolicy-TLS13-1-2-2021-06',
          certificateArn: primaryCert.arn,
          defaultActions: [
            {
              type: 'forward',
              targetGroupArn: primaryTargetGroup.arn,
            },
          ],
        },
        { provider: primaryProvider, parent: this }
      );

      // HTTPS Listener for DR ALB
      new aws.lb.Listener(
        `dr-alb-https-listener-${environmentSuffix}`,
        {
          loadBalancerArn: drAlb.arn,
          port: 443,
          protocol: 'HTTPS',
          sslPolicy: 'ELBSecurityPolicy-TLS13-1-2-2021-06',
          certificateArn: drCert.arn,
          defaultActions: [
            {
              type: 'forward',
              targetGroupArn: drTargetGroup.arn,
            },
          ],
        },
        { provider: drProvider, parent: this }
      );

      // HTTP Listener for primary ALB with redirect to HTTPS
      new aws.lb.Listener(
        `primary-alb-http-listener-${environmentSuffix}`,
        {
          loadBalancerArn: primaryAlb.arn,
          port: 80,
          protocol: 'HTTP',
          defaultActions: [
            {
              type: 'redirect',
              redirect: {
                port: '443',
                protocol: 'HTTPS',
                statusCode: 'HTTP_301',
              },
            },
          ],
        },
        { provider: primaryProvider, parent: this }
      );

      // HTTP Listener for DR ALB with redirect to HTTPS
      new aws.lb.Listener(
        `dr-alb-http-listener-${environmentSuffix}`,
        {
          loadBalancerArn: drAlb.arn,
          port: 80,
          protocol: 'HTTP',
          defaultActions: [
            {
              type: 'redirect',
              redirect: {
                port: '443',
                protocol: 'HTTPS',
                statusCode: 'HTTP_301',
              },
            },
          ],
        },
        { provider: drProvider, parent: this }
      );
    } else {
      // HTTP-only listeners for test/PR environments
      new aws.lb.Listener(
        `primary-alb-http-listener-${environmentSuffix}`,
        {
          loadBalancerArn: primaryAlb.arn,
          port: 80,
          protocol: 'HTTP',
          defaultActions: [
            {
              type: 'forward',
              targetGroupArn: primaryTargetGroup.arn,
            },
          ],
        },
        { provider: primaryProvider, parent: this }
      );

      new aws.lb.Listener(
        `dr-alb-http-listener-${environmentSuffix}`,
        {
          loadBalancerArn: drAlb.arn,
          port: 80,
          protocol: 'HTTP',
          defaultActions: [
            {
              type: 'forward',
              targetGroupArn: drTargetGroup.arn,
            },
          ],
        },
        { provider: drProvider, parent: this }
      );
    }

    this.primaryAlbDnsName = primaryAlb.dnsName;
    this.drAlbDnsName = drAlb.dnsName;
    this.primaryAlbZoneId = primaryAlb.zoneId;
    this.drAlbZoneId = drAlb.zoneId;
    this.primaryAlbArn = primaryAlb.arn;
    this.drAlbArn = drAlb.arn;
    this.primaryInstanceSecurityGroupId = primaryInstanceSg.id;
    this.drInstanceSecurityGroupId = drInstanceSg.id;

    this.registerOutputs({
      primaryAlbDnsName: this.primaryAlbDnsName,
      drAlbDnsName: this.drAlbDnsName,
      primaryAlbZoneId: this.primaryAlbZoneId,
      drAlbZoneId: this.drAlbZoneId,
      primaryAlbArn: this.primaryAlbArn,
      drAlbArn: this.drAlbArn,
      primaryInstanceSecurityGroupId: this.primaryInstanceSecurityGroupId,
      drInstanceSecurityGroupId: this.drInstanceSecurityGroupId,
    });
  }
}
